doc: release-notes-latest: added NCS specific extensions to Zephyr

tejlmand opened the pull request:
This commit updates the release notes with the NCS specific extensions
to Zephyr.

Extensions added:
- NCS Partition manager
- NCS network extensions
- MGTT extension
- Cryptocell extension

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

ru-fu commented on doc/nrf/releases/release-notes-latest.rst:
```diff
@@ -93,4 +93,11 @@ The following list summarizes the most important changes inherited from upstream
 The following list contains |NCS| specific additions:
 
-*
+* The Bluetooth controllers default uses split architecture
```
What does this mean? What is "split architecture" - split between Host and Controller?
And why "controllers" - doesn't Zephyr have only one Controller?

Maybe something like:
```suggestion
* Configured the Zephyr Bluetooth stack to build only the Host or only the Controller instead of doing a combined build (see :ref:`zephyr:bluetooth-build-types`).
  The reason for this default configuration is to add support for the :ref:`ble_controller`.
```
The links point to https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/zephyr/guides/bluetooth/bluetooth-arch.html#build-types and https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrfxlib/ble_controller/README.html#nrf-bluetooth-le-controller , respectively.

tejlmand replied:
That should actually be described in the docs, although that section seems a bit empty:
https://docs.zephyrproject.org/latest/guides/bluetooth/bluetooth-arch.html#split

tejlmand replied:
and here:
http://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/kconfig/CONFIG_BT_LL_SW_SPLIT.html#cmdoption-arg-config-bt-ll-sw-split

joerchan replied:
What is the purpose here? To describe ways that NCS is different from Zephyr? In that case I think until the SoftDevice controller is the default this paragraph can be removed.

@rugeGerritsen Any opinions here?

joerchan replied:
Note that Legacy (as opposed "split") has been removed, so Zephyr only has one choice symbol here.

@cvinayak @carlescufi https://docs.zephyrproject.org/latest/guides/bluetooth/bluetooth-arch.html#split
This could perhaps be remove as well? Only one option now.

ru-fu replied:
> What is the purpose here? To describe ways that NCS is different from Zephyr?

The idea is to point out persistent differences between NCS's Zephyr fork and upstream Zephyr.

joerchan replied:
Right, so remove this sentence then. Since this is not different for upstream.

carlescufi replied:
Yes, please remove this @tejlmand

tejlmand replied:
Then why do we still have this path ?
https://github.com/nrfconnect/sdk-zephyr/commit/e85c858c21752b50b5b9e6f26cf78300c168a882

joerchan replied:
@tejlmand Because we add another option to that menu, and since order matters we need the default statement there to keep the same default.

tejlmand replied:
got it.
Thanks for explaining :)

ru-fu commented on doc/nrf/releases/release-notes-latest.rst:
```diff
@@ -93,4 +93,11 @@ The following list summarizes the most important changes inherited from upstream
 
-*
+* The Bluetooth controllers default uses split architecture
+* NCS Partition Manager (PM) for flash partitioning
```
```suggestion
* Added support for the |NCS|'s :ref:`partition_manager`, which can be used for flash partitioning.
```

ru-fu commented on doc/nrf/releases/release-notes-latest.rst:
```diff
@@ -93,4 +93,11 @@ The following list summarizes the most important changes inherited from upstream
-*
+* The Bluetooth controllers default uses split architecture
+* NCS Partition Manager (PM) for flash partitioning
+* NCS network socket and address extensions:
+  AF_LTE, NPROTO_AT, NPROTO_PDN, NPROTO_DFU, SOCK_MGMT, SO_RCVTIMEO, SO_BINDTODEVICE, SOL_PDN, SOL_DFU, SO_PDN_CONTEXT_ID, SO_PDN_STATE, SOL_DFU, SO_DFU_ERROR, TLS_SESSION_CACHE,
+  SO_SNDTIMEO, MSG_TRUNC
+  SO_SILENCE_ALL, SO_IP_ECHO_REPLY, SO_IPV6_ECHO_REPLY
```
```suggestion
* Added the following network socket and address extensions to the :ref:`zephyr:bsd_sockets_interface` interface to support the functionality provided by the :ref:`nrfxlib:bsdlib`:
  AF_LTE, NPROTO_AT, NPROTO_PDN, NPROTO_DFU, SOCK_MGMT, SO_RCVTIMEO, SO_BINDTODEVICE, SOL_PDN, SOL_DFU, SO_PDN_CONTEXT_ID, SO_PDN_STATE, SOL_DFU, SO_DFU_ERROR, TLS_SESSION_CACHE, SO_SNDTIMEO, MSG_TRUNC, SO_SILENCE_ALL, SO_IP_ECHO_REPLY, SO_IPV6_ECHO_REPLY
```
Is this what they are needed for?

tejlmand replied:
Those are the exact extensions added.
Listing them here makes it much easier to search for them, if they are not already know to the reader.
(some are nordic specific)

mbolivar-nordic replied:
>  Listing them here makes it much easier to search for them, if they are not already know to the reader.

But how can we avoid this going stale?

ru-fu commented on doc/nrf/releases/release-notes-latest.rst:
```diff
@@ -93,4 +93,11 @@ The following list summarizes the most important changes inherited from upstream
+  AF_LTE, NPROTO_AT, NPROTO_PDN, NPROTO_DFU, SOCK_MGMT, SO_RCVTIMEO, SO_BINDTODEVICE, SOL_PDN, SOL_DFU, SO_PDN_CONTEXT_ID, SO_PDN_STATE, SOL_DFU, SO_DFU_ERROR, TLS_SESSION_CACHE,
+  SO_SNDTIMEO, MSG_TRUNC
+  SO_SILENCE_ALL, SO_IP_ECHO_REPLY, SO_IPV6_ECHO_REPLY
+* MQTT: Support enabling of TLS caching
```
```suggestion
* Added support for enabling TLS caching when using the :ref:`zephyr:mqtt_socket_interface` library.
```
How do you enable it?

tejlmand replied:
added  link.

ru-fu commented on doc/nrf/releases/release-notes-latest.rst:
```diff
@@ -93,4 +93,11 @@ The following list summarizes the most important changes inherited from upstream
+  SO_SNDTIMEO, MSG_TRUNC
+  SO_SILENCE_ALL, SO_IP_ECHO_REPLY, SO_IPV6_ECHO_REPLY
+* MQTT: Support enabling of TLS caching
+* CC310: Cryptocell access from non-secure code.
```
?
Where was that added? What is the actual change - support for CryptoCell? Or support for access from non-secure code?

tejlmand replied:
Support for access from non-secure code:
https://github.com/nrfconnect/sdk-zephyr/commit/32c5d847e8d0196e9b9a32413c67cc98089b1ef2

ru-fu replied:
```suggestion
* Updated the nrf9160ns DTS to support accessing the CryptoCell CC310 hardware from non-secure code.
```
Does that make any sense?

tejlmand commented:
> We need to add more information here. Remember that the target audience for this list is users that want to know what has changed and how they can use it - not developers who have reviewed all the PRs and know exactly what is going on in both Zephyr and NCS.

Well, isn't that the case for most entries in the release notes ?
It's a list of what has changed, not how to use is, just like for example Bluetooth:
```
Bluetooth:

-  Added support for LE Advertising Extensions.
-  Added APIs for application-controlled data length and PHY updates.
-  Added legacy OOB pairing support.
```
how to use it should be described in the general documentation.

ru-fu commented:
> > We need to add more information here. Remember that the target audience for this list is users that want to know what has changed and how they can use it - not developers who have reviewed all the PRs and know exactly what is going on in both Zephyr and NCS.
> 
> Well, isn't that the case for most entries in the release notes ?
> It's a list of what has changed, not how to use is, just like for example Bluetooth:
> 
> ```
> Bluetooth:
> 
> -  Added support for LE Advertising Extensions.
> -  Added APIs for application-controlled data length and PHY updates.
> -  Added legacy OOB pairing support.
> ```
> 
> how to use it should be described in the general documentation.

I'm not saying that we should put the full documentation here. But we need to make the entries clear enough that users know where to find that change. In the NCS release notes, we always put links to the respective libraries and samples - here we don't always do this since the list you are quoting is Zephyr content, not nRF content, and the documentation must therefore be handled by Zephyr (plus, this part will eventually be replaced by the Zephyr release notes anyway).
Your section is about the NCS-specific additions, and we need to give enough information for users to figure out what we're talking about. And I feel that if I cannot figure out how to reword something or where to link to, we're missing information.

NordicBuilder commented:
**Some checks failed. Please fix and resubmit.**

## Gitlint issues

Commit e0a6852334:
1: UC2 Body does not contain a 'Signed-off-by:' line
1: UC3 Title does not follow [subsystem]: [subject] (and should not start with literal subsys:): "Update doc/nrf/releases/release-notes-latest.rst"

## Identity/Emails issues

e0a6852334835160e311adefa1690c729b794490: author email (Torsten Tejlmand Rasmussen <tejlmand@users.noreply.github.com>) must be a real email and cannot end in @users.noreply.github.com



*Tip: The bot edits this comment instead of posting a new one, so you can check the comment's history to see earlier messages.*

tejlmand commented:
> I'm not saying that we should put the full documentation here. But we need to make the entries clear enough that users know where to find that change. 

agree, and from your actual suggestions, I also see that is the case :)

ru-fu commented on doc/nrf/releases/release-notes-latest.rst:
```diff
@@ -93,4 +93,8 @@ The following list summarizes the most important changes inherited from upstream
+* Added support for the |NCS|'s :ref:`partition_manager`, which can be used for flash partitioning.
+* Added the following network socket and address extensions to the :ref:`zephyr:bsd_sockets_interface` interface to support the functionality provided by the :ref:`nrfxlib:bsdlib`:
+  AF_LTE, NPROTO_AT, NPROTO_PDN, NPROTO_DFU, SOCK_MGMT, SO_RCVTIMEO, SO_BINDTODEVICE, SOL_PDN, SOL_DFU, SO_PDN_CONTEXT_ID, SO_PDN_STATE, SOL_DFU, SO_DFU_ERROR, TLS_SESSION_CACHE, SO_SNDTIMEO, MSG_TRUNC, SO_SILENCE_ALL, SO_IP_ECHO_REPLY, SO_IPV6_ECHO_REPLY
+* Added support for enabling TLS caching when using the :ref:`zephyr:mqtt_socket_interface` library. See also `Enabling and testing TLS in mqtt_simple`_
```
Please put the second sentence on a new line.

That link explains how to enable TLS, but is this what we mean with "TLS caching"? Then we should maybe word it differently, because it sounds like this is something different than what is described in the blog post. Or is there a new Kconfig option (or some other configuration) for TLS caching?

tejlmand replied:
> That link explains how to enable TLS, but is this what we mean with "TLS caching"?

Actually not. I just browsed the code.
I can only see that the option is now possible, but I don't find any place where we actual use this feature.
Maybe we should leave it out until we actual have a sample that shows how this can be used.
@jtguggedal any input on this ?

Will remove link again, as it only describes TLS itself and not the caching option.

jtguggedal replied:
TLS caching option has been added to the downstream version of the MQTT library (no support in native network stack, so can't be added upstream yet), so its use is shown there:
https://github.com/nrfconnect/sdk-zephyr/blob/26216d1f7e09e9c43dbcfaf0a8565bb0f69d0130/subsys/net/lib/mqtt/mqtt_transport_socket_tls.c#L81

I have a new library in the making that uses the socket option, and it will hopefully feature in NCS v1.4.0.
I can also add TLS caching to the MQTT sample in time for NCS 1.4.0 to show how it can be used.

tejlmand replied:
> TLS caching option has been added to the downstream version of the MQTT library (no support in native network stack, so can't be added upstream yet), so its use is shown there:

Well, that is also the code I found, but that just checks if the config enables it, and then sets the sock option accordingly.

What I referred to was somewhere that a sample was doing, for examlple
```
tls_config->session_cache = TLS_SESSION_CACHE_ENABLED;
```
similar to:
https://github.com/nrfconnect/sdk-nrf/blob/master/samples/nrf9160/aws_fota/src/main.c#L431-L438

and I don't find such example anywhere.
Therefore the question, if we should leave it out.

> I have a new library in the making that uses the socket option, and it will hopefully feature in NCS v1.4.0.
> I can also add TLS caching to the MQTT sample in time for NCS 1.4.0 to show how it can be used.

When that is available, then we should probably update release notes, and thus leave it out for now?

jtguggedal replied:
If it's practice to leave things out if it's not shown how to use it in a sample, I'm fine with that.
We will highlight TLS caching more in v1.4.0 and make sure it goes into the release notes then.

ru-fu replied:
There's no common practice since this is the first time we're adding these release notes. ;)
But that option is actually documented, so how about:
```suggestion
* Added support for enabling TLS caching when using the :ref:`zephyr:mqtt_socket_interface` library.
  See :c:macro:`TLS_SESSION_CACHE`.
```
This will then link to https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/zephyr/reference/networking/sockets.html?highlight=tls_session_cache#c.TLS_SESSION_CACHE

tejlmand replied:
Done

ru-fu commented on doc/nrf/releases/release-notes-latest.rst:
```diff
@@ -201,4 +201,8 @@ Other:
+* Added support for the |NCS|'s :ref:`partition_manager`, which can be used for flash partitioning.
+* Added the following network socket and address extensions to the :ref:`zephyr:bsd_sockets_interface` interface to support the functionality provided by the :ref:`nrfxlib:bsdlib`:
+  AF_LTE, NPROTO_AT, NPROTO_PDN, NPROTO_DFU, SOCK_MGMT, SO_RCVTIMEO, SO_BINDTODEVICE, SOL_PDN, SOL_DFU, SO_PDN_CONTEXT_ID, SO_PDN_STATE, SOL_DFU, SO_DFU_ERROR, TLS_SESSION_CACHE, SO_SNDTIMEO, MSG_TRUNC, SO_SILENCE_ALL, SO_IP_ECHO_REPLY, SO_IPV6_ECHO_REPLY
+* Added support for enabling TLS caching when using the :ref:`zephyr:mqtt_socket_interface` library.
```
```suggestion
* Added support for enabling TLS caching when using the :ref:`zephyr:mqtt_socket_interface` library.
  See :c:macro:`TLS_SESSION_CACHE`.
```

mbolivar-nordic commented on doc/nrf/releases/release-notes-latest.rst:
```diff
@@ -201,4 +201,9 @@ Other:
-*
+* Added support for the |NCS|'s :ref:`partition_manager`, which can be used for flash partitioning.
+* Added the following network socket and address extensions to the :ref:`zephyr:bsd_sockets_interface` interface to support the functionality provided by the :ref:`nrfxlib:bsdlib`:
+  AF_LTE, NPROTO_AT, NPROTO_PDN, NPROTO_DFU, SOCK_MGMT, SO_RCVTIMEO, SO_BINDTODEVICE, SOL_PDN, SOL_DFU, SO_PDN_CONTEXT_ID, SO_PDN_STATE, SOL_DFU, SO_DFU_ERROR, TLS_SESSION_CACHE, SO_SNDTIMEO, MSG_TRUNC, SO_SILENCE_ALL, SO_IP_ECHO_REPLY, SO_IPV6_ECHO_REPLY
```
I think we need some documentation added to our internal synchronization procedures to keep this list updated at each upmerge too.

tejlmand replied:
agree.
